File: Examples/HandTracking/CPP/usecases_ht.cpp

#include "stdafx.h"

#include "usecases_ht.h"

// std::vector<Vec3<Real>> to_pose_params(Vec<Real> const& theta, size_t n_bones)
// {
//   size_t n = 3 + n_bones;
//   std::vector<Vec3<Real>> pose_params{n, vec(0.0,0.0,0.0) };
//   pose_params[0] = theta.segment<3>(0);
//   pose_params[1] = vec(1.0, 1.0, 1.0);
//   pose_params[2] = theta.segment<3>(3);

//   size_t i_theta = 6;
//   int n_fingers = 5;
//   for (int i = 0; i < n_fingers; ++i) {
//     pose_params[3 + i * 3 + 0][0] = theta[i_theta++];
//     pose_params[3 + i * 3 + 0][1] = theta[i_theta++];
//     pose_params[3 + i * 3 + 1][0] = theta[i_theta++];
//     pose_params[3 + i * 3 + 2][0] = theta[i_theta++];

//   }
//   return pose_params;
// }


// Real3x3 euler_angles_to_rotation_matrix(Vec3<Real> xzy)
// {
//   auto tx = xzy[0];
//   auto tz = xzy[1];
//   auto ty = xzy[2];

//   Real3x3 Rx = vec(
//     vec(1., 0., 0.),
//     vec(0., cos(tx), -sin(tx)),
//     vec(0., sin(tx), cos(tx)));

//   Real3x3 Ry = vec(
//     vec(cos(ty), 0., sin(ty)),
//     vec(0., 1., 0.),
//     vec(-sin(ty), 0., cos(ty)));

//   Real3x3 Rz = vec(
//     vec(cos(tz), -sin(tz), 0.),
//     vec(sin(tz), cos(tz), 0.),
//     vec(0., 0., 1.));

//   return Rz*Ry*Rx;
// }

// Real4x4 Rt_to_transform(Real3x3 const& R, Real3 const& t)
// {
//   Mat<Real, 4, 4> T;
//   for (int i = 0; i < 3; ++i) {
//     for (int j = 0; j < 3; ++j)
//       T(i, j) = R(i, j);
//     T(i, 3) = T(3, i) = 0.0;
//   }
//   T(3, 3) = 1.0;
//   return T;
// }

// auto make_relative(Vec3<Real> const& xzy, Real4x4 const& base_relative)
// {
//   auto R = euler_angles_to_rotation_matrix(xzy);
//   Mat<Real, 4, 4> T;
//   for (int i = 0; i < 3; ++i) {
//     for (int j = 0; j < 3; ++j)
//       T(i, j) = R(i, j);
//     T(i, 3) = T(3, i) = 0.0;
//   }
//   T(3, 3) = 1.0;
//   return base_relative * T;
// }


// auto get_posed_relatives(std::vector<Real4x4> const& base_relatives, std::vector<Vec3<Real>> pose_params)
// {
//   size_t offset = 3;
//   size_t n_bones = base_relatives.size();

//   std::vector<Real4x4> ret{ n_bones };
//   for (size_t i_bone = 0; i_bone < n_bones; ++i_bone)
//     ret[i_bone] = make_relative(pose_params[i_bone + offset], base_relatives[i_bone]);

//   return ret;
// }

// auto relatives_to_absolutes(std::vector<Real4x4> const& relatives, std::vector<int> const& parents)
// {
//   std::vector<Real4x4> absolutes{ relatives.size(), identity<Real,4>() };
//   for (size_t i = 0; i < parents.size(); ++i)
//     if (parents[i] == -1)
//       absolutes[i] = relatives[i];
//     else
//       absolutes[i] = absolutes[parents[i]] * relatives[i];
//   return absolutes;
// }

Real3x3 angle_axis_to_rotation_matrix(Real3 const& angle_axis)
{
  auto n = sqrt(sumsq(angle_axis));
  if (n < 0.0001)
    return vec(
      vec(1., 0., 0.),
      vec(0., 1., 0.),
      vec(0., 0., 1.)
      );

  Real x = angle_axis[0] / n;
  Real y = angle_axis[1] / n;
  Real z = angle_axis[2] / n;

  Real s = sin(n);
  Real c = cos(n);

  return vec(
    vec(x*x + (1. - x*x)*c, x*y*(1. - c) - z*s, x*z*(1. - c) + y*s),
    vec(x*y*(1. - c) + z*s, y*y + (1. - y*y)*c, y*z*(1. - c) - x*s),
    vec(x*z*(1. - c) - y*s, z*y*(1. - c) + x*s, z*z + (1. - z*z)*c));
}

// auto apply_global_transform(std::vector<Vec3<Real>> const& pose_params, Mat<Real, 3, 0> const& positions)
// {
//   auto R = angle_axis_to_rotation_matrix(pose_params[0]);
//   Real3x3 scale = diaginv(pose_params[1]);
//   R = scale * R;
//   auto ret = R * positions;
//   for (size_t i = 0; i < positions.size(); ++i)
//     ret[i] = positions[i] + pose_params[2];
//   return ret;
// }

// Mat<Real, 3, N_VERTS> 
// get_skinned_vertex_positions(std::vector<Real4x4> const& base_relatives,
//   std::vector<int> const& parents,
//   Mat<Real, 3, N_VERTS> const& base_positions,
//   Mat<Real, N_VERTS, N_BONES> const& weights,
//   std::vector<Vec3<Real>> const& pose_params)
// {
//   auto relatives = get_posed_relatives(base_relatives, pose_params);
//   auto absolutes = relatives_to_absolutes(relatives, parents);

//   auto transforms = absolutes; //  Array.map2 (*) absolutes model.inverse_base_absolutes

//   size_t n_verts = base_positions.cols();

//   Mat<Real, 4, N_VERTS> base_positions_homg{ 4, n_verts, 0.0 };
//   for (size_t i = 0; i < n_verts; ++i)
//     base_positions_homg[i] = vec(base_positions[i][0], base_positions[i][1], base_positions[i][2], 1.0);

//   Mat<Real, 3, 0> positions{ 3, n_verts, 0.0 };

//   for (size_t i_transform = 0; i_transform < transforms.size(); ++i_transform) {
//     Mat<Real, 3, N_VERTS> curr_positions = transforms[i_transform].block<3,4>(0,0) * base_positions_homg;
//     for (int j = 0; j < n_verts; ++j)
//       positions[j] = positions[j] + curr_positions[j].segment<3>(0) * weights(j, i_transform);
//   }

//   return apply_global_transform(pose_params, positions);
// }

File: Examples/HandTracking/FSmooth/usecases_ht.fs
[<ReflectedDefinition>]
module usecases_ht

open corelang
open linalg
open types
open utils
open cardinality

(** Hand Tracking **)

let matrix3DUpdate (m: Matrix[]) (s: Index) (e: Index) (nm: Matrix[]): Matrix[] = 
  build (length m) (fun i ->
    let isInRange = i >= s && i < e
    if isInRange then
      nm.[i - s]
    else
      m.[i]
  )

let inline to_pose_params (theta: Vector) (n_bones: Cardinality): Matrix =
  let row1 = theta.[0..2]
  let row2 = [| 1.0; 1.0; 1.0|]
  let row3 = theta.[3..5]
  let zeroRow = [| 0.0; 0.0; 0.0 |]
  let pose_params = [| row1; row2; row3; zeroRow; zeroRow |]
  (* TODO rewrite using fold *) 
  let i1 = 5
  let finger1 = 
    [| [| theta.[i1]; theta.[i1+1]; 0.0 |] ; 
       [| theta.[i1+2]; 0.0; 0.0 |] ;
       [| theta.[i1+3]; 0.0; 0.0 |] ;
       [| 0.0; 0.0; 0.0 |] |]
  let i2 = i1 + 4
  let finger2 = 
    [| [| theta.[i2]; theta.[i2+1]; 0.0 |] ; 
       [| theta.[i2+2]; 0.0; 0.0 |] ;
       [| theta.[i2+3]; 0.0; 0.0 |] ;
       [| 0.0; 0.0; 0.0 |] |]
  let i3 = i2 + 4
  let finger3 = 
    [| [| theta.[i3]; theta.[i3+1]; 0.0 |] ; 
       [| theta.[i3+2]; 0.0; 0.0 |] ;
       [| theta.[i3+3]; 0.0; 0.0 |] ;
       [| 0.0; 0.0; 0.0 |] |]
  let i4 = i3 + 4
  let finger4 = 
    [| [| theta.[i4]; theta.[i4+1]; 0.0 |] ; 
       [| theta.[i4+2]; 0.0; 0.0 |] ;
       [| theta.[i4+3]; 0.0; 0.0 |] ;
       [| 0.0; 0.0; 0.0 |] |]
  let i5 = i4 + 4
  let finger5 = 
    [| [| theta.[i5]; theta.[i5+1]; 0.0 |] ; 
       [| theta.[i5+2]; 0.0; 0.0 |] ;
       [| theta.[i5+3]; 0.0; 0.0 |] ;
       [| 0.0; 0.0; 0.0 |] |]

  matrixConcat pose_params 
    (matrixConcat finger1 
      (matrixConcat finger2 
        (matrixConcat finger3 
          (matrixConcat finger4 finger5))))

let euler_angles_to_rotation_matrix (xzy: Vector): Matrix =
  let tx = xzy.[0]
  let ty = xzy.[2]
  let tz = xzy.[1]
  let Rx = [| [|1.; 0.; 0.|]; [|0.; cos(tx); -sin(tx)|]; [|0.; sin(tx); cos(tx)|] |]
  let Ry = [| [|cos(ty); 0.; sin(ty)|]; [|0.; 1.; 0.|]; [|-sin(ty); 0.; cos(ty)|] |]
  let Rz = [| [|cos(tz); -sin(tz); 0.|]; [|sin(tz); cos(tz); 0.|]; [|0.; 0.; 1.|] |]
  matrixMult Rz (matrixMult Ry Rx)

[<DontInline>]
let make_relative (pose_params: Vector) (base_relative: Matrix): Matrix =
  let R = euler_angles_to_rotation_matrix pose_params
  let r1 = [| [| 0. |]; [| 0. |]; [| 0. |] |]
  let r2 = [| [| 0.; 0.; 0.; 1.0|] |]
  let T = 
    matrixConcat (matrixConcatCol R 
                    (r1)) 
                 (r2)
  matrixMult base_relative T

let get_posed_relatives (n_bones: Cardinality) (pose_params: Matrix) (base_relatives: Matrix[]): Matrix[] =
  let offset = 3
  vectorMapToMatrix3D (fun i_bone -> 
     make_relative pose_params.[(int i_bone)+offset] base_relatives.[int i_bone]
    ) 
    (vectorRange (Card 0) (n_bones .- (Card 1)))

let angle_axis_to_rotation_matrix (angle_axis: Vector): Matrix =
  let n = sqrt(sqnorm angle_axis)
  if n < 0.0001 then
    [| [| 1.; 0.; 0. |];
       [| 0.; 1.; 0. |];
       [| 0.; 0.; 1. |]; |]
  else
    let x = angle_axis.[0] / n
    let y = angle_axis.[1] / n
    let z = angle_axis.[2] / n

    let s = sin n
    let c = cos n
    
    [| [| x*x + (1. - x*x)*c; x*y*(1. - c) - z*s; x*z*(1. - c) + y*s |]; 
       [| x*y*(1. - c) + z*s; y*y + (1. - y*y)*c; y*z*(1. - c) - x*s |];
       [| x*z*(1. - c) - y*s; z*y*(1. - c) + x*s; z*z + (1. - z*z)*c |] |]

let relatives_to_absolutes (relatives: Matrix[]) (parents: Vector): Matrix[] =
  // let init = [| [| [| |] |] |]
  // iterateMatrix3D (fun acc i ->
  //   if parents.[i] = -1.0 then 
  //     (* Revealed a bug in ANF convertor and let lifter. Inlining the next let binding makes the code generator crash. *)
  //     let newMatrix = [| relatives.[i] |]
  //     matrix3DConcat acc newMatrix
  //   else 
  //     (* Revealed a bug in ANF convertor and let lifter. Inlining the next let binding makes the code generator crash. *)
  //     let newMatrix = [| matrixMult acc.[int parents.[i]] relatives.[i] |]
  //     matrix3DConcat acc newMatrix
  // ) (init) (Card 0) ((length relatives) .- (Card 1))
  let init = relatives
  foldOnRange (fun acc i ->
    if parents.[i] = -1.0 then 
      (* Revealed a bug in ANF convertor and let lifter. Inlining the next let binding makes the code generator crash. *)
      let newMatrix = [| relatives.[i] |]
      matrix3DUpdate acc i (i + 1) newMatrix
    else 
      (* Revealed a bug in ANF convertor and let lifter. Inlining the next let binding makes the code generator crash. *)
      let newMatrix = [| matrixMult acc.[int parents.[i]] relatives.[i] |]
      matrix3DUpdate acc i (i + 1) newMatrix
  ) (init) (Card 0) (length relatives)

let apply_global_transform (pose_params: Matrix) (positions: Matrix) = 
  let R = angle_axis_to_rotation_matrix pose_params.[0]
  let scale = pose_params.[1] // 1 row vector
  let R1 = matrixMap (fun row -> mult_vec_elementwise row scale) R 
  
  let T = matrixConcatCol R1 (matrixTranspose ([| pose_params.[2] |]))
  let ones = vectorMap (fun x -> 1.) (vectorRange (Card 1) (length positions.[0]))
  let positions_homog = matrixConcat positions ([| ones |])
  matrixMult T positions_homog

let get_skinned_vertex_positions (is_mirrored: Index) (n_bones: Cardinality) (pose_params: Matrix) (base_relatives: Matrix[]) (parents: Vector)
     (inverse_base_absolutes: Matrix[]) (base_positions: Matrix) (weights: Matrix) =
  let relatives = get_posed_relatives n_bones pose_params base_relatives
  let absolutes = relatives_to_absolutes relatives parents
  
  let transforms = matrix3DMap2 (matrixMult) absolutes inverse_base_absolutes
  
  let n_verts = length base_positions.[0]
  let init_positions = matrixFill (Card 3) n_verts 0.

  let positions = 
    foldOnRange (fun acc i_transform ->
      let curr_positions = matrixMult transforms.[i_transform].[0..2] base_positions
      let w = matrixFillFromVector (length base_positions) weights.[i_transform]
      matrixAdd acc (matrixMultElementwise curr_positions w)
    ) init_positions (Card 0) (length transforms)

  let mirrored_positions =
    if(is_mirrored = 1) then 
      let mirror_matrix = 
        [| [| -1.; 0.; 0. |];
           [| 0.; 1.; 0.  |];
           [| 0.; 0.; 1.  |] |]
      matrixMult mirror_matrix positions
    else 
      positions

  apply_global_transform pose_params mirrored_positions

let hand_objective (is_mirrored: Index) (param: Vector) (correspondences: Vector) (points: Matrix)
      (n_bones: Cardinality) (base_relatives: Matrix[]) (parents: Vector)
      (inverse_base_absolutes: Matrix[]) (base_positions: Matrix) (weights: Matrix): Vector =
  let pose_params = to_pose_params param n_bones
  
  let vertex_positions = 
    get_skinned_vertex_positions is_mirrored n_bones pose_params base_relatives parents
      inverse_base_absolutes base_positions weights
  
  
  let n_corr_card = length correspondences
  let n_corr = cardToInt n_corr_card
  let err = 
    vectorMap (fun i ->
      let ind = int i
      let r = ind / n_corr
      let c = ind % n_corr
      points.[r].[c] - vertex_positions.[r].[int correspondences.[c]]
    ) (vectorRange (Card 0) (n_corr_card .+ n_corr_card .+ n_corr_card .- (Card 1)))
  err

(** Testing **)

let test_ht () =
  let a = [| 1.0; 2.0; 3.0 |]
  vectorPrint a
  let mat1 = 
    [| [| 1.0; 2.0; 3.0; |];
       [| 4.0; 5.0; 6.0; |];
       [| 7.0; 8.0; 9.0; |] |]
  let base_rel = 
    build (Card 4) (fun i ->
      build (Card 4) (fun j ->
        double ((i+1) * 4 + j)
      )
    )
  let q = make_relative a base_rel
  matrixPrint q
  let r = angle_axis_to_rotation_matrix a
  matrixPrint r
  let s = apply_global_transform mat1 mat1
  matrixPrint s
  ()